lib/src/client/http2_connection.dart

```diff
@@ -426,6 +456,51 @@ class _SocketTransportConnector implements ClientTransportConnector {
     if (validator == null) return false;
     return validator(certificate, authority);
   }
+
+  Future<Stream<List<int>>> connectToProxy() async {
+    final headers = {'Host': '$_host:$_port'};
+    if (proxy.isAuthenticated) {
+      // If the proxy configuration contains user information use that
+      // for proxy basic authorization.
+      final authStr = '${proxy.username}:${proxy.password}';
+      final auth = base64Encode(utf8.encode(authStr));
+      headers[HttpHeaders.proxyAuthorizationHeader] = 'Basic $auth';
+    }
+    final completer = Completer<void>();
+
+    /// Routes the events through after connection to the proxy has been
+    /// established.
+    final intermediate = StreamController<List<int>>();
+
+    socket.listen(
+      (event) {
+        if (completer.isCompleted) {
+          intermediate.sink.add(event);
+        } else {
+          waitFor(event, completer);
+        }
+      },
+      onDone: intermediate.close,
+      onError: intermediate.addError,
+    );
+
+    _sendConnect(headers);
+    await completer.future;
+    return intermediate.stream;
+  }
+
+  void waitFor(Uint8List event, Completer<void> completer) {
+    final response = ascii.decode(event);
+    final lines = response.split('\r\n');
+    // status line
+    final statusLine = lines.first;
```

Comment:
What happens with the rest of data in this chunk? Is it somehow guaranteed that this chunk will only contain a single line? Needs a comment.